go/teams/transactions.go

```diff
@@ -0,0 +1,501 @@
+		Type: "keybase",
+		Name: uv.TeamInviteName(),
+		ID:   NewInviteID(),
+	}
+
+	switch role {
+	case keybase1.TeamRole_READER:
+		payload.Readers = appendToInviteList(invite, payload.Readers)
+	case keybase1.TeamRole_WRITER:
+		payload.Writers = appendToInviteList(invite, payload.Writers)
+	case keybase1.TeamRole_ADMIN:
+		payload.Admins = appendToInviteList(invite, payload.Admins)
+	case keybase1.TeamRole_OWNER:
+		payload.Owners = appendToInviteList(invite, payload.Owners)
+	default:
+		return fmt.Errorf("Unexpected role: %v", role)
+	}
+	return nil
+}
+
+// sweepMembers will queue "removes" for all cryptomembers with given
+// UID.
+func (tx *AddMemberTx) sweepMembers(uid keybase1.UID) {
+	team := tx.team
+	for chainUv := range team.chain().inner.UserLog {
+		if chainUv.Uid == uid && team.chain().getUserRole(chainUv) != keybase1.TeamRole_NONE {
+			tx.removeMember(chainUv)
+		}
+	}
+}
+
+// sweepKeybaseInvites will queue "cancels" for all keybase-type
+// invites (PUKless members) for given UID.
+func (tx *AddMemberTx) sweepKeybaseInvites(uid keybase1.UID) {
+	team := tx.team
+	for _, invite := range team.chain().inner.ActiveInvites {
+		if inviteUv, err := invite.KeybaseUserVersion(); err == nil {
+			if inviteUv.Uid == uid {
+				tx.cancelInvite(invite.Id)
+			}
+		}
+	}
+}
+
+// AddMemberByUsername will add member by username and role. It
+// checks if given username can become crypto member or a PUKless
+// member. It will also clean up old invites and memberships if
+// necessary.
+func (tx *AddMemberTx) AddMemberByUsername(ctx context.Context, username string, role keybase1.TeamRole) (err error) {
+	team := tx.team
+	g := team.G()
+
+	defer g.CTrace(ctx, fmt.Sprintf("AddMemberTx.AddMemberByUsername(%s,%v)", username, role), func() error { return err })()
+	g.Log.CDebugf(ctx, "AddMemberByUsername(%s, %v) to team %q", username, role, team.Name())
+
+	inviteRequired := false
+	normalizedUsername, uv, err := loadUserVersionPlusByUsername(ctx, g, username)
+	g.Log.CDebugf(ctx, "AddMemberByUsername: loaded user %q -> (%q, %v, %v)", username, normalizedUsername, uv, err)
+	if err != nil {
+		if err == errInviteRequired {
+			inviteRequired = true
+			g.Log.CDebugf(ctx, "Invite required for %v", uv)
+		} else {
+			return err
+		}
+	}
+
+	// Do not do partial updates here. If error is returned, it is
+	// assumed that tx is untouched, and caller can continue with
+	// other attempts. This is used in batch member adds, when even if
+	// some users can't be added, it skips them and continues with
+	// others.
+
+	if role == keybase1.TeamRole_OWNER && team.IsSubteam() {
+		return NewSubteamOwnersError()
+	}
+
+	if team.IsMember(ctx, uv) {
+		if inviteRequired {
+			return fmt.Errorf("user is already member but we got errInviteRequired")
+		}
+		return libkb.ExistsError{Msg: fmt.Sprintf("user %s is already a member of team %q",
+			normalizedUsername, team.Name())}
+	}
+
+	existingUV, err := team.UserVersionByUID(ctx, uv.Uid)
+	if err == nil {
```

Comment:
combine with the above line, so that `existingUV` is scoped just for the purposes of the `if`